setuptools/dist.py

```diff
@@ -36,9 +46,35 @@
 from distutils.fancy_getopt import translate_longopt
 from distutils.util import strtobool
 
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
 __all__ = ['Distribution']
 
-sequence = tuple, list
+_sequence = tuple, list
+"""
+:meta private:
+
+Supported iterable types that are known to be:
+- ordered (which `set` isn't)
+- not match a str (which `Sequence[str]` does)
+- not imply a nested type (like `dict`)
+for use with `isinstance`.
+"""
+_Sequence: TypeAlias = Union[Tuple[str, ...], List[str]]
+# This is how stringifying _Sequence would look in Python 3.10
+_requence_type_repr = "tuple[str, ...] | list[str]"
+
+
+def __getattr__(name: str) -> Any:  # pragma: no cover
+    if name == "sequence":
+        SetuptoolsDeprecationWarning.emit(
+            "`setuptools.dist.sequence` is an internal implementation detail.",
+            "Please define your own `sequence = tuple, list` instead.",
+            due_date=(2025, 8, 28),  # Originally added on 2024-08-27
+        )
+        return _sequence
+    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
```

Comment:
@Avasam, how about this for the deprecation of `sequence`?

After a [quick search on github](https://github.com/search?q=%2F%28from+setuptools.*sequence%29%7C%28import+setuptools.*sequence%29%2F+language%3APython&type=code), I do see people accessing it, so we cannot simply remove it...

Reply:
I had opted out of wanting to replace`sequence` by `typing.Sequence` in https://github.com/pypa/setuptools/pull/4575#discussion_r1720850125 because there's a lot of implications when it comes to new checks and supported types (like having to explicitely check for `str` first everytime, because a `str` is a valid `Sequence[str]`, both at runtime and in type-checking).
Not to say it wouldn't be an improvement, just that's not something I wanna go into for the scope of those PRs.

Is making `sequence` private something you want to do independent of those changes ?

Reply:
> Is making sequence private something you want to do independent of those changes ?

I think so. It seems to be such a trivial implementation detail, I don't know why people are importing it from setuptools.

It is probably good to deprecate it now just in case at some point we decide to remove it.